src/content/en/2021/javascript.md

```diff
@@ -1,20 +1,1543 @@
+
+[Custom Elements](https://developers.google.com/web/fundamentals/web-components/customelements) are customized elements built on top of the [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement) class. The browser provides a `customElements` API that allows developers to define an element and register it with the browser as a custom element. 
+
+It is noted that 2.7% of desktop pages use custom elements for one or more parts of the web page.
+
+{{ figure_markup(
+  content="2.7%",
+  caption="Percent of desktop pages using custom elements.",
+  classes="big-number",
+  sheets_gid="1991863136",
+  sql_file="web_components_specs.sql"
+  )
+}}
+
+[Shadow DOM](https://developers.google.com/web/fundamentals/web-components/shadowdom) allows you to create a dedicated subtree in the DOM for the custom element introduced to the browser. It ensures the styles and nodes inside the element are not accessible outside the element.
+
+0.4% of pages use Shadow DOM specification of web components to ensure a scoped subtree for the element.
+
+{{ figure_markup(
+  content="0.4%",
+  caption="Percent of pages using Shadow DOM.",
+  classes="big-number",
+  sheets_gid="1991863136",
+  sql_file="web_components_specs.sql"
+  )
+}}
+
+A [&lt;template>](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots#the_truth_about_templates) element is very useful when there is a pattern in the markup which could be reused. The contents of _template_ elements render only when referenced by JavaScript. 
+
+Templates work well when dealing with web components, as the content that is not yet referenced by JavaScript is then appended to a shadow root using the shadow DOM.
+
+
+{{ figure_markup(
+  content="<0.1%",
+  caption="Percent of pages using templates.",
+  classes="big-number",
+  sheets_gid="1991863136",
+  sql_file="web_components_specs.sql"
+  )
+}}
+
+Less than 0.1% of web pages have adopted the use of templates. Although templates are well [supported](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots#:~:text=Templates%20are%20well%2Dsupported%20in%20browsers%3B%20the%20Shadow%20DOM%20API%20is%20supported%20by%20default%20in%20Firefox%20(version%2063%20onwards)%2C%20Chrome%2C%20Opera%2C%20Safari%2C%20and%C2%A0Edge) in browsers, there is still a very low percentage of pages using them.
+
+
+### Preload and Prefetch
+
+At the render of a page, the browser downloads the given resources and to prioritise the download of some resources over others the browser uses [resource hints](https://almanac.httparchive.org/en/2021/resource-hints). The resource hints, preload and prefetch, are basically to tell the browser to either download some resource right away or to download it whenever it can to be able to present the resource when required.
+
+The **_preload_** hint tells the browser to download the resource with a higher priority as it will be required on the current page.  \
```

Comment:
```suggestion
The **_preload_** hint tells the browser to download the resource with a higher priority as it will be required on the current page.
```